### PR TITLE
Solution Issue #719

### PR DESCRIPTION
**Pull request by epoell**
[Solution to Issue to #719 ](https://github.com/eclipse/xtext-xtend/issues/719)
If a type is named Abstractfoo but not declared abstract, an issue is raised.
The severity is configurable, default is warning. Also a quick fix is available to add the missing abstract modifier. If a the type is also declared final (but named Abstractfoo), the final modifier is removed, as it can be assumed that the abstract modifier is more important to the user (the type's name takes more time and thought to type in than the final keyword).

### CODE REVIEW

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java [hunk 1]
```diff
@@ -687,6 +688,72 @@ public void apply(IModificationContext context) throws Exception {
...
+					} else {
+						//use existent method to add abstract modifier, if modifier list is empty, to prevent wrong insertion
+						internalDoAddAbstractKeyword(element, context);
+					}
+				}
+			}
+		});
+	}
+	
+	@Fix(IssueCodes.MODIFIER_DOES_NOT_MATCH_TYPENAME)
```
Comment (szarnekow) [inline org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: Looks like this method was added twice.

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java [hunk 2]
```diff
@@ -687,6 +688,72 @@ public void apply(IModificationContext context) throws Exception {
 		});
 	}
 	
+	@Fix(IssueCodes.MODIFIER_DOES_NOT_MATCH_TYPENAME) 
+	public void addModifierAccordingToTypename(final Issue issue, IssueResolutionAcceptor acceptor){
```
Comment (cdietrich) [inline org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: missing space between `)` and `{` => can you format the whole method

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java [hunk 3]
```diff
@@ -2168,7 +2168,16 @@ protected boolean isApi(XtendTypeDeclaration type) {
 		}
 		return api;
 	}
-	
+
+	@Check
+	protected void checkModifierMatchesClassname(XtendClass xtendClass) {
+		String name = xtendClass.getName();
+		if (name.startsWith("Abstract") && !xtendClass.isAbstract()) {
+			warning("Your class is named " + name + ", but not declared abstract.", XTEND_TYPE_DECLARATION__NAME, -1,
```
Comment (cdietrich) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java]: should be

```
addIssue("Your class is named " + name + ", but not declared abstract.", xtendClass, XTEND_TYPE_DECLARATION__NAME, -1,
					MODIFIER_DOES_NOT_MATCH_TYPENAME);
```

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java [hunk 4]
```diff
@@ -687,6 +688,72 @@ public void apply(IModificationContext context) throws Exception {
 		});
 	}
 	
+	@Fix(IssueCodes.MODIFIER_DOES_NOT_MATCH_TYPENAME)
```
Comment (cdietrich) [inline org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: if i call the quickfix on the first element in

```
class AbstractXxxxx{   
	static class AbstractYyyyyy1 {
		
	}
	static class AbstractYyyyyy2 {
		
	}
	static class AbstractYyyyyy3 {
		
	}
}
```
the last one gets fixed

Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: this could actually be bug in org.eclipse.xtend.ide.quickfix.XtendQuickfixProvider.internalDoAddAbstractKeyword(EObject, IModificationContext)

looks like it misses a break in the for loop once the class is found

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	def add_abstract_package() {
+		val input = '''
+			package class AbstractSample {
+			}
+		'''
+		internTestQuickFixPackage(input)
+	}
+
+	@Test
+	def add_abstract_package_final() {
```
Comment (cdietrich) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: could we add tests with multiples classes + class with inner class too (see bug hint on the quickfix provider)

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ * Copyright (c) 2019 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtend.core.tests.validation
+
+import org.eclipse.xtend.core.tests.AbstractXtendTestCase
+import javax.inject.Inject
```
Comment (miklossy) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend]: should be `com.google.inject.Inject`

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import org.junit.Test
+import org.eclipse.xtext.testing.util.ParseHelper
+import org.eclipse.xtext.testing.validation.ValidationTestHelper
+import org.eclipse.xtend.core.xtend.XtendFile
+import static org.eclipse.xtend.core.xtend.XtendPackage.Literals.*
+import static org.eclipse.xtend.core.validation.IssueCodes.*
```
Comment (miklossy) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend]: Organize imports

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @author Eva P�ll - Initial contribution
+ */
+class ValidationIssue719Test extends AbstractXtendTestCase {
+	@Inject extension ValidationTestHelper
+	@Inject extension ParseHelper<XtendFile>
+
+	@Test
+	def void namedAndDeclaredAbstract() {
```
Comment (miklossy) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend]: remove the explicit `void` return type, they are unnecessary

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		file.assertNoIssues
+	}
+
+	@Test
+	def void namedAndNotDeclaredAbstract() {
+		val file = '''
+			class AbstractC {
+			}
+		'''.parse
+		file.assertWarning(XTEND_CLASS, MODIFIER_DOES_NOT_MATCH_TYPENAME, "not declared abstract");
```
Comment (miklossy) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend]: Remove unnecessary `;`

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	
+	@Test
+	def void namedAndNotDeclaredAbstractPackage() {
+		val file = '''
+			package class AbstractC {
+			}
+		'''.parse
+		file.assertWarning(XTEND_CLASS, MODIFIER_DOES_NOT_MATCH_TYPENAME, "not declared abstract");
+	}
+
```
Comment (miklossy) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend]: I would prefer a more compact implementation, e.g: 
	```
@Test def namedAndNotDeclaredAbstract() {
		'''
			class AbstractC {
			}
		'''.parse.assertWarning(XTEND_CLASS, MODIFIER_DOES_NOT_MATCH_TYPENAME, "not declared abstract")
	}
```

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend (hunk 8 above)
Comment (cdietrich) [reply org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/ValidationIssue719Test.xtend]: i rather prefer them on tests. @szarnekow what do you think?

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendConfigurableIssueCodes.java [hunk 11]
```diff
@@ -8,6 +8,8 @@
 package org.eclipse.xtend.core.validation;
 
 
+import static org.eclipse.xtend.core.validation.IssueCodes.*;
```
Comment (miklossy) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendConfigurableIssueCodes.java]: Unused import.

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import static extension org.eclipse.xtext.ui.testing.util.JavaProjectSetupUtil.createJavaProject
+
+/**
+ * @author epoell - Initial contribution and API
+ */
+@RunWith(XtextRunner)
+@InjectWith(XtendIDEInjectorProvider)
+class ValidationIssue719QuickFixTest extends AbstractQuickfixTest {
+
+	@Inject extension QuickfixTestBuilder builder
```
Comment (miklossy) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: The `QuickfixTestBuilder` should be not necessary here, please remove.

Comment (miklossy) [reply org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: Actually, I found the entire QuickfixTest test case very hard to read. I would prefer a more readable implementation, something like:
```
/**
 * @author epoell - Initial contribution and API
 */
@RunWith(XtextRunner)
@InjectWith(XtendIDEInjectorProvider)
class ValidationIssue719QuickFixTest extends AbstractQuickfixTest {

	@Before def void setup() {
		/*
		 * Xbase-based languages require java project
		 */
		projectName.createJavaProject
	}

	@Test def add_abstract_simple() {
		'''
			class AbstractSample {
			}
		'''.testAbstractModifierQuickfix('''
			abstract class AbstractSample {
			}
		''')
	}

	@Test def add_abstract_final_simple() {
		'''
			final class AbstractSample {
			}
		'''.testAbstractModifierQuickfix('''
			abstract class AbstractSample {
			}
		''')
	}

	@Test def add_abstract_package() {
		'''
			package class AbstractSample {
			}
		'''.testAbstractModifierQuickfix('''
			package abstract class AbstractSample {
			}
		''')
	}

	@Test def add_abstract_package_final() {
		'''
			package final class AbstractSample {
			}
		'''.testAbstractModifierQuickfix('''
			package abstract class AbstractSample {
			}
		''')
	}

	private def testAbstractModifierQuickfix(CharSequence input, String result) {
		val issueCode = MODIFIER_DOES_NOT_MATCH_TYPENAME
		val label = "Add missing abstract modifier."
		val description = "Add the abstract modifier to match naming conventions for the type name. Delete final modifier if necessary."
		input.testQuickfixesOn(issueCode, new Quickfix(label, description, result))
	}
```

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java (hunk 2 above)
Comment (epoell) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: Can I just use the eclipse formatter in general, or is there some kind of "how to format nicely" sheet?

Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: assuming you have created the environment with oomph, marking the method and do a normal format in eclipse should do the trick

Comment (epoell) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: alright, thanks :+1:

**Comment by cdietrich** [comment]
@epoell
please have a look at the compile errors

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.3.0:compile (default-compile) on project org.eclipse.xtend.ide: Compilation failure: Compilation failure: 
[ERROR] /jobs/genie.xtext/xtext-xtend/branches/PR-724/workspace/org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java:[692] 
[ERROR] 	public void addModifierAccordingToTypename(final Issue issue, IssueResolutionAcceptor acceptor) {
[ERROR] 	            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] Duplicate method addModifierAccordingToTypename(Issue, IssueResolutionAcceptor) in type XtendQuickfixProvider
[ERROR] /jobs/genie.xtext/xtext-xtend/branches/PR-724/workspace/org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java:[725] 
[ERROR] 	public void addModifierAccordingToTypename(final Issue issue, IssueResolutionAcceptor acceptor){
[ERROR] 	            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] Duplicate method addModifierAccordingToTypename(Issue, IssueResolutionAcceptor) in type XtendQuickfixProvider
[ERROR] /jobs/genie.xtext/xtext-xtend/branches/PR-724/workspace/org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java:[758] 
[ERROR] 	public void addModifierAccordingToTypename(final Issue issue, IssueResolutionAcceptor acceptor){
[ERROR] 	            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java [hunk 13]
```diff
@@ -687,6 +688,106 @@ public void apply(IModificationContext context) throws Exception {
...
+	public void addModifierAccordingToTypename(final Issue issue, IssueResolutionAcceptor acceptor) {
```
Comment (cdietrich) [inline org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: @epoell method seems to be here 3 times

Comment (epoell) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: I look into it, it seems I have not tamed Git for me, yet. Sorry!

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java (hunk 4 above)
Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: this problem is still open

**Comment by cdietrich** [comment]
we still should have a look at the Quickfix in this usecase

```
class AbstractXxxxx{   
	static class AbstractYyyyyy1 {
		
	}
	static class AbstractYyyyyy2 {
		
	}
	static class AbstractYyyyyy3 {
		
	}
}
```
still believe

`org.eclipse.xtend.ide.quickfix.XtendQuickfixProvider.internalDoAddAbstractKeyword(EObject, IModificationContext)`

needs to `break` the for loop once the first class keyword is found

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java (hunk 13 above)
Comment (epoell) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: It might work now

**Comment by epoell** [comment]
I could look into it now, if my git trials finally work.

**Comment by epoell** [comment]
a `break` within the for-loop of `internalDoAddAbstractKeyword` seems to solve the problem.
I'm writing some testcases first and will commit it afterwards.

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java (hunk 4 above)
Comment (epoell) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: This bug also affected the quickfixes for `CLASS_MUST_BE_ABSTRACT` and `MISSING_ABSTRACT`.
So, if only the signature of a method is given and there is no implemented method in the class or if a method is declared abstract, but the class is not, the wrong insertion happened, too.

    class Sample2 {
    
        static class Innerclass {
        }
    
        def void foo()
        def abstract int abstractFoo()
    }

Here the other two quickfixes would have inserted abstract at `Innerclass` That is fixed by the `break` statement, too.

Should the tests for these quickfixes be extended, too, to mirror this bug?

Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: it would be ok for me to test it once

**Comment by epoell** [comment]
I think, I'm ready :)

- well with one question pending: So far I marked all Tests as returning void (explicitly). Shall I leave or change that?

**Comment by miklossy** [comment]
You can leave it as is.

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+	@Test
+	def void add_abstract_nested_classes_outer() {
+		'''
+			class AbstractX{
+				abstract static class AbstractY{
+				}
+			}
+		'''.testQuickfixesOn(MODIFIER_DOES_NOT_MATCH_TYPENAME,
+			new Quickfix("Add missing abstract modifier.",
```
Comment (miklossy) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: why not a call on the internTestQuickFix private method?

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java [hunk 15]
```diff
@@ -2168,7 +2168,16 @@ protected boolean isApi(XtendTypeDeclaration type) {
 		}
 		return api;
 	}
-	
+
+	@Check
+	protected void checkModifierMatchesTypename(XtendClass xtendClass) {
+		String name = xtendClass.getName();
+		if (name.startsWith("Abstract") && !xtendClass.isAbstract()) {
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java]: The name could be null with incomplete models. Please guard this.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java [hunk 16]
```diff
@@ -2168,7 +2168,16 @@ protected boolean isApi(XtendTypeDeclaration type) {
...
+			addIssue("Your type is named " + name + ", but not declared abstract.", xtendClass, XTEND_TYPE_DECLARATION__NAME, -1,
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java]: I'm not convinced yet. Can we rephrase this to 'The class <name> is not declared abstract.'

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+	@Before
+	def void setup() {
+		// Xbase-based languages require java project
+		projectName.createJavaProject
+	}
+
+	@Test
+	def void add_abstract_simple() {
+		'''
+			class AbstractSample {
```
Comment (szarnekow) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: Could you please add a test where the class has an annotation and another one where it has a JavaDoc comment?

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+	@Test
+	def void add_abstract_nested_classes_outer() {
+		'''
+			class AbstractX{
+				abstract static class AbstractY{
+				}
+			}
+		'''.internTestQuickFix('''
+					abstract class AbstractX{
```
Comment (szarnekow) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: Indentation looks weird

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+			}
+		'''.internTestQuickFix('''
+			abstract static class AbstractX{
+				class SampleInner{
+				}
+			}
+		''')
+	}
+
+	def private internTestQuickFix(CharSequence input, CharSequence result) {
```
Comment (szarnekow) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/ValidationIssue719QuickFixTest.xtend]: I'm not super convinced by the method name. Either `internalTestQuickFix` or simply `applyFix`

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java [hunk 20]
```diff
@@ -687,6 +688,40 @@ public void apply(IModificationContext context) throws Exception {
...
+					modifiers.remove("final");
+
+					//add abstract
+					if (modifiers.contains("public") || modifiers.contains("package")) {
+						index++;
+					}
+					if (modExist) {
+						modifiers.add(index, "abstract");
+					} else {
+						//use existent method to add abstract modifier, if modifier list is empty, to prevent wrong insertion
```
Comment (szarnekow) [inline org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/quickfix/XtendQuickfixProvider.java]: existent -> existing? Also it's more a remark that is obvious: if there was no existing method, it could not have been used. 
Maybe simplify to `// workaround for bug xyz - fallback to textual modification

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java (hunk 15 above)
Comment (epoell) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/XtendValidator.java]: Is it sufficient to check if the name is not null, or should also an exception be thrown?
